src/platform/wayland.rs

```diff
@@ -73,9 +80,22 @@ impl EventLoopBuilderExtWayland for EventLoopBuilder {
 /// Additional methods on [`Window`] that are specific to Wayland.
 ///
 /// [`Window`]: crate::window::Window
-pub trait WindowExtWayland {}
+pub trait WindowExtWayland {
+    /// Returns `xdg_toplevel` of the window or [`None`] if the window is X11 window.
+    fn xdg_toplevel(&self) -> Option<NonNull<c_void>>;
+}
 
-impl WindowExtWayland for dyn CoreWindow + '_ {}
+#[cfg(wayland_platform)]
+impl WindowExtWayland for dyn CoreWindow + '_ {
+    #[inline]
+    fn xdg_toplevel(&self) -> Option<NonNull<c_void>> {
+        let w = self.as_any().downcast_ref::<crate::platform_impl::wayland::Window>()?;
+        let id = w.xdg_toplevel().id();
+        let ptr = NonNull::new(id.as_ptr().cast()).expect("xdg_toplevel should not be null");
```

Comment:
could you move this logic to `wayland/window/mod.rs` and return pointer from there instead, so we don't have an export here. We could also not fail when we can not `cast`.

Reply:
Done.

> We could also not fail when we can not `cast`.

Sorry I don't understand what you mean. Do you mean we should use `NonNull::new_unchecked` instead of `NonNull::new`?

Reply:
I mean, that you just forward option up and not panic.

Reply:
Done.